File: frontend/src/types/scan.ts
export type ScanStatus = 'pending' | 'running' | 'completed' | 'failed'

export type FindingSeverity = 'low' | 'medium' | 'high'
export type ScanExportFormat = 'json' | 'csv'
export type DiscoveredLinksResponse = {
  scan_id: string
  target: string
  links: string[]
}

export type ScanFinding = {
  id: string
  category: string
  severity: FindingSeverity
  title: string
  description: string
  recommendation: string
  evidence: Record<string, unknown>
  created_at: string
}

export type ScanSummary = {
  id: string
  target: string
  normalized_target: string
  target_type: 'domain' | 'ip'
  status: ScanStatus
  score: number | null
  summary: string | null
  started_at: string | null
  completed_at: string | null
  created_at: string
  updated_at: string
}

export type ReportSnapshot = {
  id: string
  http_headers: Record<string, unknown>
  tls_analysis: Record<string, unknown>
  port_results: Array<Record<string, unknown>>
  misconfigurations: Array<Record<string, unknown>>
  metadata: Record<string, unknown>
  created_at: string
}

export type ScanDetail = ScanSummary & {
  findings: ScanFinding[]
  snapshot: ReportSnapshot | null
}

File: backend/app/schemas/scan.py
from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field


class ScanCreateRequest(BaseModel):
    target: str


class FindingResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    category: str
    severity: str
    title: str
    description: str
    recommendation: str
    evidence: dict[str, object]
    created_at: datetime


class ReportSnapshotResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    http_headers: dict[str, object]
    tls_analysis: dict[str, object]
    port_results: list[dict[str, object]]
    misconfigurations: list[dict[str, object]]
    metadata: dict[str, object] = Field(validation_alias="report_metadata")
    created_at: datetime


class ScanListItemResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    target: str
    normalized_target: str
    target_type: str
    status: str
    score: int | None
    summary: str | None
    started_at: datetime | None
    completed_at: datetime | None
    created_at: datetime
    updated_at: datetime


class ScanDetailResponse(ScanListItemResponse):
    findings: list[FindingResponse]
    snapshot: ReportSnapshotResponse | None
